### PR TITLE
Create build and update traffic during backends:create

### PR DESCRIPTION
**Pull request by taeold**
After creating a build via the "backends:create" command, follow up by creating a new build and updating the traffic setting to the newly created build.

This allows users to access their newly created site without having to manually trigger a rollout themselves.

### CODE REVIEW

File: src/gcp/apphosting.ts [hunk 1]
```diff
@@ -46,33 +47,139 @@ export interface Build {
...
+  minInstances?: number;
+  memory?: string;
+}
 
 interface BuildSource {
-  codeBaseSource?: CodebaseSource;
+  codebase: CodebaseSource;
 }
 
+interface CodebaseSource {
```
Comment (tonyjhuang) [inline src/gcp/apphosting.ts:87]: Pretty brutal we can't reuse protos here :/

File: src/init/features/apphosting/index.ts [hunk 2]
```diff
@@ -66,27 +72,50 @@ export async function doSetup(setup: any, projectId: string): Promise<void> {
...
+      { children: [build.error] }
+    );
+  }
+
+  await setDefaultTraffic(projectId, location, backendId, build.name.split("/").pop()!);
 
   if (backend) {
     logSuccess(`Successfully created backend:\n\t${backend.name}`);
-    logSuccess(`Your site is being deployed at:\n\thttps://${backend.uri}`);
+    logSuccess(`Your site is now deployed at:\n\thttps://${backend.uri}`);
```
Comment (tonyjhuang) [inline src/init/features/apphosting/index.ts:102]: why change?

File: src/init/features/apphosting/index.ts [hunk 3]
```diff
@@ -66,27 +72,50 @@ export async function doSetup(setup: any, projectId: string): Promise<void> {
...
+  });
+
+  if (build.state !== "READY") {
+    throw new FirebaseError(
+      `Failed to build your app. Please inspect the build logs at ${build.buildLogsUri}.`,
+      { children: [build.error] }
+    );
+  }
+
+  await setDefaultTraffic(projectId, location, backendId, build.name.split("/").pop()!);
```
Comment (tonyjhuang) [inline src/init/features/apphosting/index.ts:98]: please add a line comment with the expected format of build.name

File: src/init/features/apphosting/index.ts [hunk 4]
```diff
@@ -136,6 +157,69 @@ export async function createBackend(
...
+ */
+function generateBuildId(n = 6) {
+  const letters = "abcdefghijklmnopqrstuvwxyz";
+  const allChars = "01234567890-abcdefghijklmnopqrstuvwxyz";
+  let id = letters[Math.floor(Math.random() * letters.length)];
+  for (let i = 1; i < n; i++) {
+    const idx = Math.floor(Math.random() * allChars.length);
+    id += allChars[idx];
+  }
+  return id;
```
Comment (tonyjhuang) [inline src/init/features/apphosting/index.ts:174]: uuid?

File: src/init/features/apphosting/index.ts [hunk 5]
```diff
@@ -136,6 +157,69 @@ export async function createBackend(
...
+ * Create (and waits for) a new build.
+ */
+export async function createBuild(
+  projectId: string,
+  location: string,
+  backendId: string,
+  buildInput: Omit<BuildInput, "name">
+): Promise<Build> {
+  logBullet("Creating a new build... this may take a few minutes.");
+  const buildId = generateBuildId();
```
Comment (tonyjhuang) [inline src/init/features/apphosting/index.ts:187]: is build id a required field? Can we let the server create it for us?

**Review by tonyjhuang — APPROVED** [review]
tests? :)

File: src/init/features/apphosting/index.ts (hunk 2 above)
Comment (taeold) [reply src/init/features/apphosting/index.ts]: Becaused at this point, the site would've have been deployed!

File: src/init/features/apphosting/index.ts (hunk 3 above)
Comment (taeold) [reply src/init/features/apphosting/index.ts]: Had to refactor a bit, and this function no longer exists.

File: src/gcp/apphosting.ts (hunk 1 above)
Comment (taeold) [reply src/gcp/apphosting.ts:87]: I would've love to use AI to generate the types 😢

File: src/init/features/apphosting/index.ts (hunk 4 above)
Comment (taeold) [reply src/init/features/apphosting/index.ts]: that could work except i don't think uuid's are guaranteed to start with non-number

File: src/init/features/apphosting/index.ts (hunk 5 above)
Comment (taeold) [reply src/init/features/apphosting/index.ts]: Unfortunately, yes for now.

**Comment by taeold** [comment]
@tonyjhuang going to add some tech debt here since testing UI flow is a bit brittle. We'll need to refactor so that various helper functions are easily testable.

**Comment by taeold** [comment]
### Merge activity

* **Dec 13, 8:06 PM**: [Graphite](https://app.graphite.dev/github/pr/firebase/firebase-tools/6611) rebased this pull request after merging its parent, because this pull request is set to merge when ready.

**Pull request merged**